### PR TITLE
* Start work on `VisualStudio` themes

### PR DESCRIPTION
**Pull request by Wagnerp**
* More work

image

### CODE REVIEW

File: Source/Krypton Components/Krypton.Ribbon/General/CalculatedValues.cs [hunk 1]
```diff
@@ -79,6 +79,7 @@ public void Recalculate()
                     case PaletteRibbonShape.Microsoft365:
                     case PaletteRibbonShape.VisualStudio:
                     case PaletteRibbonShape.Office2010:
+                    case PaletteRibbonShape.VisualStudio2010:
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Ribbon/General/CalculatedValues.cs:82]: I thought that I saw the palette called vs20**22**

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 2]
```diff
@@ -806,13 +833,89 @@ protected PaletteBase()
...
         /// <param name="style">Style of button spec.</param>
         /// <param name="state">State for which image is required.</param>
         /// <returns>Image value.</returns>
-        public abstract Image? GetButtonSpecImage(PaletteButtonSpecStyle style, PaletteState state);
+        public virtual Image? GetButtonSpecImage(PaletteButtonSpecStyle style, PaletteState state)
+        {
+            switch (style)
+            {
+                case PaletteButtonSpecStyle.Close:
+                    return ReturnButtonSpecImageArray()[0];
```
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:848]: > Can the storage be a map of the enum to the image ?

I've made a `TODO` comment reminding me to do this, just wanted to ensure that the idea works first.

File: Source/Krypton Components/Krypton.Toolkit/Rendering/RenderDefinitions.cs [hunk 3]
```diff
@@ -878,6 +878,11 @@ public enum RendererMode
         /// </summary>
         Microsoft365,
 
+        /// <summary>
+        /// Specifies the RenderVisualStudio be used.
+        /// </summary>
+        VisualStudio,
```
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Rendering/RenderDefinitions.cs:884]: There's going to be two values, `VisualStudio2010` for the 2010 style (before it became flat) & `VisualStudio` which will cover 2012 - 2022 colours.

File: Source/Krypton Components/Krypton.Ribbon/General/CalculatedValues.cs (hunk 1 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Ribbon/General/CalculatedValues.cs:82]: The `VisualStudio2010` will be for the 2010 style (which will be based off O2k10) & `VisualStudio` will be based off the MS365 themes with colour variations between 2012 - 2022.

File: Documents/Help/Changelog.md [hunk 4]
```diff
@@ -3,6 +3,7 @@
 =======
 
 ## 2023-11-xx - Build 2311 - November 2023
+* Resolved [#1063](https://github.com/Krypton-Suite/Standard-Toolkit/issues/1063), `ToolTipValues` not found in `KryptonContextMenuItem`
```
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Documents/Help/Changelog.md:6]: Removed

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 5]
```diff
@@ -92,7 +93,33 @@ protected PaletteBase()
             // Inherit means we need to calculate the value next time it is requested
             _dragFeedback = PaletteDragFeedback.Inherit;
 
-            _themeName = null;
+            _themeName = string.Empty;
+
+            _useKryptonFileDialogs = false;
```
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:98]: Now true

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs [hunk 6]
```diff
@@ -253,7 +254,47 @@ public abstract class PaletteMicrosoft365Base : PaletteBase
         private string _baseFontName;
         #endregion
 
-        #region Constructor        
+        #region Identity
+
+        static PaletteMicrosoft365Base()
+        {
+            _buttonSpecImages = new Image[]
```
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:261]: @Smurf-IV Can I use `BiDictionary` as seen in `ThemeManager`?

File: Documents/Help/Changelog.md [hunk 7]
```diff
@@ -3,7 +3,6 @@
 =======
 
 ## 2023-11-xx - Build 2311 - November 2023
-* Resolved [#1063](https://github.com/Krypton-Suite/Standard-Toolkit/issues/1063), `ToolTipValues` not found in `KryptonContextMenuItem`
 * Implemented [#1015](https://github.com/Krypton-Suite/Standard-Toolkit/issues/1057), Show a tooltips on a `KryptonContextMenuItem`
```
Comment (Smurf-IV) [inline Documents/Help/Changelog.md]: Link number is incorrect.

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs [hunk 8]
```diff
@@ -0,0 +1,18 @@
+﻿#region BSD License
+/*
+ * 
+ *  New BSD 3-Clause License (https://github.com/Krypton-Suite/Standard-Toolkit/blob/master/LICENSE)
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs:4]: If this is a new file, then the License header needs to change

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 9]
```diff
@@ -26,7 +26,7 @@ public abstract class PaletteBase : Component
         private Padding? _inputControlPadding;
         private PaletteDragFeedback _dragFeedback;
         private string _themeName;
-        private Image[] /*ICollection<Image?>*/ _buttonSpecImages;
+        protected static Image[] /*ICollection<Image?>*/ _buttonSpecImages;
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:29]: What happens when a new palette is loaded, how will these images be replaced ?
i.e. Not sure static is the right option.

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs (hunk 9 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:29]: Yes, I noticed while testing today that it doesn't work. Going to revert.

Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:29]: Threw a assertion

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:348]: Going to use `IDictionary<PaletteButtonSpecStyle, Image>`

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Visual Studio/Official Themes/2012/PaletteVisualStudio2012DarkMode.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Visual Studio/Official Themes/2012/PaletteVisualStudio2012DarkMode.cs:1]: New files need license and the removal of the `using`'s to follow the same pattern as the other files

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs [hunk 11]
```diff
@@ -168,6 +166,46 @@ private static readonly Padding
 
         #endregion
 
+        #region Image Array
+
+        private static readonly Image[] __buttonSpecImages = new Image[]
+        {
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:172]: These will still need magic numbers to access (Ok, they may be enums' but what if one of these palettes introduces an image in the middle?)

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs (hunk 11 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:172]: I think that a 'key & value' solution would be better.

File: Documents/Help/Changelog.md (hunk 7 above)
Comment (Wagnerp) [reply Documents/Help/Changelog.md]: Removed link

File: Source/Krypton Components/Krypton.Toolkit/Controls Toolkit/KryptonCustomPaletteBase.cs [hunk 12]
```diff
@@ -402,6 +402,82 @@ public KryptonPaletteCheckButtons ButtonStyles
 
         #endregion
 
+        #region Image Dictionary
+
+        protected override IDictionary<PaletteButtonSpecStyle, Image> ButtonSpecStyleImageDictionary()
+        {
+            IDictionary<PaletteButtonSpecStyle, Image> imageDictionary =
+                new ConcurrentDictionary<PaletteButtonSpecStyle, Image>();
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Controls Toolkit/KryptonCustomPaletteBase.cs:410]: Does not need to be a concurrent dictionary, as this will be mostly for reading, which is thread safe anyway. Concurrent is for multithreaded Modifications (add / remove)

Comment (Smurf-IV) [reply Source/Krypton Components/Krypton.Toolkit/Controls Toolkit/KryptonCustomPaletteBase.cs:410]: Also every time this is called, it will be creating a new dictionary of images. This is going to be slow and wasteful.

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 13]
```diff
@@ -127,6 +128,33 @@ protected PaletteBase([DisallowNull] Image[] /*ICollection<Image>*/ buttonSpecIm
...
+            // Inherit means we need to calculate the value next time it is requested
+            _dragFeedback = PaletteDragFeedback.Inherit;
+
+            _themeName = string.Empty;
+
+            _useKryptonFileDialogs = true;
+
+            _baseFont = _defaultFontStyle;
+
+            _baseFontSize = 9f;
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:155]: Why is this hardcoded, It should be from the default style

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs (hunk 8 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs:4]: Will be deleting this file, found another way.

Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs:4]: On second thought, keeping this file, but changing licence header.

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs (hunk 13 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs]: Now dynamically set

File: Source/Krypton Components/Krypton.Toolkit/Controls Toolkit/KryptonCustomPaletteBase.cs (hunk 12 above)
Comment (Wagnerp) [reply Source/Krypton Components/Krypton.Toolkit/Controls Toolkit/KryptonCustomPaletteBase.cs]: Reworked, so the image dictionary is set up in the palettes

**Comment by Wagnerp** [comment]
@Smurf-IV Not sure where I'm going wrong, but the values in `IntegratedToolBarValues` don't show up in KForm?

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs [hunk 14]
```diff
@@ -882,79 +837,75 @@ protected PaletteBase([DisallowNull] IDictionary<PaletteButtonSpecStyle, Image>
         /// <returns>Image value.</returns>
         public virtual Image? GetButtonSpecImage(PaletteButtonSpecStyle style, PaletteState state)
         {
-            Dictionary<PaletteButtonSpecStyle, Image> imageDictionary;
-
-            if (GetButtonSpecImageDictionary().Values != null)
-            {
-                imageDictionary = ConvertToDictionary(GetButtonSpecImageDictionary());
-            }
+            Dictionary<PaletteButtonSpecStyle, Image> imageDictionary =
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Base/PaletteBase.cs:840]: Is this used anywhere in this function ?

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs [hunk 15]
```diff
@@ -3009,16 +2928,16 @@ public override int GetContentAdjacentGap(PaletteContentStyle style, PaletteStat
 
         #region Image Dictionary
 
-        protected override IDictionary<PaletteButtonSpecStyle, Image> ButtonSpecStyleImageDictionary()
+        /// <summary>Setups the button spec image dictionary.</summary>
+        private void SetupButtonSpecImageDictionary()
         {
-            IDictionary<PaletteButtonSpecStyle, Image> imageDictionary =
-                new ConcurrentDictionary<PaletteButtonSpecStyle, Image>();
+            Dictionary<PaletteButtonSpecStyle, Image> imageDictionary = new Dictionary<PaletteButtonSpecStyle, Image>();
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Bases/PaletteMicrosoft365Base.cs:2934]: Is this used ?

File: Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Extra Themes/PaletteMicrosoft365BlueLightMode.cs [hunk 16]
```diff
@@ -3774,16 +3736,16 @@ public override Padding GetMetricPadding(PaletteState state, PaletteMetricPaddin
 
         #region Image Dictionary
 
-        protected override IDictionary<PaletteButtonSpecStyle, Image> ButtonSpecStyleImageDictionary()
+        /// <summary>Setups the button spec image dictionary.</summary>
+        private void SetupButtonSpecImageDictionary()
         {
-            IDictionary<PaletteButtonSpecStyle, Image> imageDictionary =
-                new ConcurrentDictionary<PaletteButtonSpecStyle, Image>();
+            Dictionary<PaletteButtonSpecStyle, Image> imageDictionary = new Dictionary<PaletteButtonSpecStyle, Image>();
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Palette Builtin/Microsoft 365/Extra Themes/PaletteMicrosoft365BlueLightMode.cs:3742]: etc.

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

File: Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs [hunk 17]
```diff
@@ -1,7 +1,6 @@
 ﻿#region BSD License
 /*
  * 
- *  New BSD 3-Clause License (https://github.com/Krypton-Suite/Standard-Toolkit/blob/master/LICENSE)
```
Comment (Smurf-IV) [inline Source/Krypton Components/Krypton.Toolkit/Converters/ImageLookupConverter.cs]: Why was this line removed ? It's needed as part of the BSD requirement.

**Review by Smurf-IV — CHANGES_REQUESTED** [review]

**Comment by Smurf-IV** [comment]
> @Smurf-IV Not sure where I'm going wrong, but the values in `IntegratedToolBarValues` don't show up in KForm?

Fix / change the integrated stuff in a separate PR, try and keep this one for the VS theme

**Comment by Wagnerp** [comment]
@Smurf-IV Going to close, copy over the VS theme changes into a new PR
